docs/lake/design.md

```diff
@@ -0,0 +1,657 @@
+present in the query.
+
+While an individual search lookup involves latency to cloud storage to lookup
+a key in each index, each lookup is cheap and involves a small amount of data
+and the lookups can all be run in parallel, even from a single node, so
+the scan schedule can be quickly computed in a small number of round-trips
+(that navigate very wide B-trees) to cloud object storage.
+
+### Commit Journal
+
+The pool's journal is the definitive record of the evolution of data in
+that pool.  At any point in the journal, a snapshot may be available
+for efficient, pool-key range scans.
+
+Each journal entry is identified with its `journal ID`,
+a 64-bit, unsigned integer that begins at 0.
+The journal may be updated concurrently by multiple writers so concurrency
+controls are included (see below) to provide atomic updates.
+
+A journal entry simply contains one or more ADD and DROP directives,
+which refer to a commit tag and its key range, which in turn, implies
+one or more data segments that comprise said commit.  A journal entry
+also contains descriptive information as described earlier.
+
+Because each reference commit includes its key range and because the
+commit tag embeds a time stamp (indicating date/time of creation),
+journal entries can be purged with _either_ key-based or time-based
+retention policies (or both).
+
+Each journal is a ZNG file numbered 0 to the end of journal (HEAD),
+e.g., `0.zng`, `1.zng`, etc.
+The journal's TAIL begins at 0 and is increased as journal entries are purged.
+Entries are added at the HEAD and removed from the TAIL.
+Once created, a journal entry is never modified.
+
+Each journal entry implies a snapshot of the data in a pool.  A snapshot
+is computed by applying the ADD/DROP directives in sequence from entry TAIL to
+the journal entry in question, up to HEAD.  This gives the set of commit tags
+that comprise a snapshot.  A snapshot may then be scanned by scanning,
+in key order, the segments that comprise all of the commits while merging
+records from overlapping segments.  The snapshot is sorted by its pool key
+range, where key-range values are sorted by the beginning key as the primary key
+and the ending key is the secondary key.
+
+For efficiency, a journal entry's snapshot may be stored as "cached snapshot"
+alongside the journal entry.  This way, the snapshot at HEAD may be
+efficiently computed by locating the most recent cached snapshot and scanning
+forward to HEAD.
+
+#### Scaling the Journal
+
+When the size of the snapshot file reaches a certain size (and thus becomes too large to
+conveniently handle in memory), the journal is converted to an internal sub-pool
+called the "journal pool".  The journal pool's
+pool key is the "from" value (of its parent pool key) from each commit.
+In this case, commits to the parent pool are made in the same fashion,
+but instead of snapshotting updates into a snapshot zng file,
```

Comment:
```suggestion
but instead of snapshotting updates into a snapshot ZNG file,
```

You've been pretty consistent in here about capitalizing ZNG when speaking about it in sentences (which I support), so, might as well keep it up. :smile: